zebra-state/src/service/chain_tip.rs

```diff
@@ -181,3 +201,133 @@ impl ChainTip for ChainTipReceiver {
             .unwrap_or_else(|| Arc::new([]))
     }
 }
+
+/// A chain tip change monitor.
+/// Used to `await` changes and resets of the state's best chain tip.
+///
+/// The latest changes are available from all cloned instances of this type,
+/// but each clone separately tracks the last change it provided,
+/// so it can provide a [`Reset`] if the best chain fork changes.
+///
+/// The chain tip data is based on:
+/// * the best non-finalized chain tip, if available, or
+/// * the finalized tip.
+#[derive(Debug)]
+pub struct ChainTipChange {
+    /// The receiver for the current chain tip's data.
+    receiver: watch::Receiver<ChainTipData>,
+
+    /// The most recent hash provided by this instance.
+    previous_change_hash: Option<block::Hash>,
+}
+
+/// Actions that we can take in response to a [`ChainTipChange`].
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum TipAction {
+    /// The chain tip was updated continuously,
+    /// using a child `block` of the previous block.
+    ///
+    /// The genesis block action is a `Grow`.
+    Grow {
+        /// Information about the block used to grow the chain.
+        block: ChainTipBlock,
+    },
+
+    /// The chain tip was reset to a block with `height` and `hash`.
+    ///
+    /// Resets can happen for different reasons:
+    /// * a newly created or cloned [`ChainTipChange`], which is behind the current tip,
+    /// * extending the chain with a network upgrade activation block,
+    /// * switching to a different best [`Chain`], also known as a rollback, and
+    /// * receiving multiple blocks since the previous change.
+    ///
+    /// To keep the code and tests simple, Zebra performs the same reset actions,
+    /// regardless of the reset reason.
+    ///
+    /// `Reset`s do not have the transaction hashes from the tip block,
+    /// because all transactions should be cleared by a reset.
+    Reset {
+        /// The block height of the tip, after the chain reset.
+        height: block::Height,
+
+        /// The block hash of the tip, after the chain reset.
+        ///
+        /// Mainly useful for logging and debugging.
+        hash: block::Hash,
+    },
+}
+
+impl ChainTipChange {
+    /// Wait until the next chain tip change, then return the corresponding [`TipAction`].
+    ///
+    /// Returns an error if communication with the state is lost.
+    pub async fn next(&mut self) -> Result<TipAction, watch::error::RecvError> {
+        let block = self.next_block().await?;
+
+        // TODO: handle resets here
```

Comment:
Question: is this going to be handled in this PR or in a separate PR?

Reply:
In a separate PR

Reply:
See PR #2721